File: feign-api/src/main/java/name/abhijitsarkar/feign/persistence/RecordingRequest.java
/*
 * Copyright (c) 2016, the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * License for more details.
 */

package name.abhijitsarkar.feign.persistence;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.ToString;
import name.abhijitsarkar.feign.model.Request;
import org.springframework.data.annotation.Id;

import java.util.Objects;

/**
 * @author Abhijit Sarkar
 */
@NoArgsConstructor
@ToString
@SuppressWarnings({"PMD.BeanMembersShouldSerialize", "PMD.ImmutableField", "PMD.SingularField"})
public class RecordingRequest extends Request {
    @Id
    @Getter
    private String id;

    public RecordingRequest(Request request, String id) {
        super();
        path = request.getPath();
        method = request.getMethod();
        queryParams = request.getQueryParams();
        headers = request.getHeaders();
        body = request.getBody();

        this.id = id;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        RecordingRequest that = (RecordingRequest) o;
        return Objects.equals(id, that.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
}

File: feign-api/src/main/scala/name/abhijitsarkar/feign/persistence/RecordingRequest.scala
/*
 * Copyright (c) 2016, the original author or authors.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 3 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * A copy of the GNU General Public License accompanies this software,
 * and is also available at http://www.gnu.org/licenses.
 */

package name.abhijitsarkar.feign.persistence

import java.util.{Collections, Objects, Optional, List => JavaList, Map => JavaMap}

import name.abhijitsarkar.feign.Request
import org.springframework.data.annotation.Id

import scala.beans.BeanProperty
import scala.collection.JavaConverters._

/**
  * @author Abhijit Sarkar
  */
class RecordingRequest {
  @Id
  @BeanProperty
  var id: String = _
  @BeanProperty
  var path: String = "/**"
  @BeanProperty
  var method: String = "GET"
  @BeanProperty
  var queryParams: JavaMap[String, JavaList[String]] = Collections.emptyMap()
  @BeanProperty
  var headers: JavaMap[String, String] = Collections.emptyMap()
  @BeanProperty
  var body: String = _

  def copyFrom(request: Request): RecordingRequest = {
    this.path = request.path
    this.method = request.method
    this.queryParams = request.queryParams.mapValues(_.asJava).asJava
    this.headers = request.headers.asJava
    this.body = request.body.getOrElse(null)

    this
  }

  override def equals(that: Any): Boolean =
    that match {
      case t: RecordingRequest => t.id == id
      case _ => false
    }

  override def hashCode: Int = Objects.hash(id)
}